src/char/char.c

```diff
@@ -765,123 +755,196 @@ int char_memitemdata_to_sql(const struct item items[], int max, int id, int tabl
-		SQL->StmtBindColumn(stmt, 10 + MAX_SLOTS + j * 2, SQLDT_INT16, &item.option[j].index, 0, NULL, NULL);
-		SQL->StmtBindColumn(stmt, 11 + MAX_SLOTS + j * 2, SQLDT_INT16, &item.option[j].value, 0, NULL, NULL);
+
+	for (i = 0; i < MAX_SLOTS; i++)
+		SQL->StmtBindColumn(stmt, 10 + i, SQLDT_SHORT, &item.card[i], 0, NULL, NULL);
+
+	for (i = 0; i < MAX_ITEM_OPTIONS; i++) {
+		SQL->StmtBindColumn(stmt, 10 + MAX_SLOTS + i * 2, SQLDT_INT16, &item.option[i].index, 0, NULL, NULL);
+		SQL->StmtBindColumn(stmt, 11 + MAX_SLOTS + i * 2, SQLDT_INT16, &item.option[i].value, 0, NULL, NULL);
 	}
+
 	if (has_favorite)
 		SQL->StmtBindColumn(stmt, 10 + MAX_SLOTS + MAX_ITEM_OPTIONS * 2, SQLDT_UCHAR, &item.favorite, 0, NULL, NULL);
 
-	// bit array indicating which inventory items have already been matched
-	flag = aCalloc(max, sizeof(bool));
+	if (SQL->StmtNumRows(stmt) > 0 ) {
+		i = 0;
+		while (SQL_SUCCESS == SQL->StmtNextRow(stmt) && i < max) {
+			memcpy(&items[i], &item, sizeof(struct item));
+			i++;
+		}
+	}
+
+	return i;
+}
 
-	while (SQL_SUCCESS == SQL->StmtNextRow(stmt)) {
-		found = false;
-		// search for the presence of the item in the char's inventory
-		for (i = 0; i < max; ++i) {
-			// skip empty and already matched entries
-			if (items[i].nameid == 0 || flag[i])
+/**
+ * Saves an array of 'item' entries into the specified table. [Smokexyz/Hercules]
+ * @param items       The items array.
+ * @param max         The array size.
+ * @param guid        The character/account/guild ID (depending on table).
+ * @param tableswitch The type of table (@see enum inventory_table_type).
+ * @retval -1 in case of failure, or number of changes made within the table.
+ */
+int char_memitemdata_to_sql(const struct item *p_items, int max, int guid, enum inventory_table_type table)
+{
+	StringBuf buf;
+	int i = 0, j = 0;
+	const char *tablename = NULL;
+	const char *selectoption = NULL;
+	bool has_favorite = false;
+	struct item *cp_items = NULL; // temp item storage variable
+	bool *matched_p = NULL;
+	int total_updates = 0, total_deletes = 0, total_inserts = 0, total_changes = 0;
+	int *deletes = NULL;
+
+	nullpo_ret(p_items);
+	Assert_retr(1, max > 0); // Anything below wouldn't work without a max limit.
+
+	switch (table) {
+	case TABLE_INVENTORY:     tablename = inventory_db;     selectoption = "char_id";    has_favorite = true; break;
+	case TABLE_CART:          tablename = cart_db;          selectoption = "char_id";    break;
+	case TABLE_GUILD_STORAGE: tablename = guild_storage_db; selectoption = "guild_id";   break;
+	default:
+		ShowError("Invalid table name!\n");
+		Assert_retr(-1, table);
+	}
+
+	cp_items = aCalloc(max, sizeof(struct item));
+	deletes = aCalloc(max, sizeof(struct item));
+	matched_p = aCalloc(max, sizeof(bool));
+
+	StrBuf->Init(&buf);
+
+	/**
+	 * If the storage table is not empty, check for items and replace or delete where needed.
+	 */
+	if (chr->getitemdata_from_sql(cp_items, max, guid, table) > 0) {
+		for (i = 0; i < max; i++) {
```

Comment:
Optimization: `chr->getitemdata_from_sql()` returns the actual data size, so there's no need to loop through the trailing empty values until `max` here.